sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,270 @@
+	defer cancel()
+
+	setRecordingMode("live")
+	err := p.GlobalSetup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runSetup(p PerfTest) error {
+	fmt.Println("Running `Setup`")
+
+	ctx, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.Setup(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// runTest takes care of the semantics of running a single iteration. It returns the number of times the test ran as an int, the exact number
+// of seconds the test ran as a float64, and any errors.
+func runTest(p PerfTest, c chan runResult) {
+	defer wg.Done()
+	fmt.Println("Beginning `Run`")
+	ctx, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	// If we are using the test proxy need to set up the in-memory recording.
+	if TestProxy != "" {
+		// First request goes through in Live mode
+		setRecordingMode("live")
+		err := p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+
+		// 2nd request goes through in Record mode
+		setRecordingMode("record")
+		start(p.GetMetadata(), nil)
+		err = p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+		stop(p.GetMetadata(), nil)
+
+		// All ensuing requests go through in Playback mode
+		setRecordingMode("playback")
+		start(p.GetMetadata(), nil)
+	}
+
+	start := time.Now()
+	count := 0
+	for time.Since(start).Seconds() < float64(Duration) {
+		err := p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+		count += 1
+	}
+
+	elapsed := time.Since(start).Seconds()
+
+	// Stop the proxy now
+	stop(p.GetMetadata(), nil)
+	setRecordingMode("live")
+	c <- runResult{count: count, timeInSeconds: elapsed, err: nil}
+}
+
+// runTearDown takes care of the semantics for tearing down a single iteration of a performance test.
+func runTearDown(p PerfTest) error {
+	fmt.Println("Running `Teardown`")
+
+	_, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.TearDown(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runGlobalTearDown(p PerfTest) error {
+	fmt.Println("Running `GlobalTeardown`")
+	_, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.GlobalTearDown(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type runResult struct {
+	count         int
+	timeInSeconds float64
+	err           error
+}
+
+func runPerfTest(p PerfTest) error {
+	err := runGlobalSetup(p)
+	if err != nil {
+		return err
+	}
+
+	totalCount := 0
+	totalTime := 0.0
+
+	runSetup(p)
+
+	var channels []chan runResult
+
+	for idx := 0; idx < Parallel; idx++ {
+		wg.Add(1)
+		c := make(chan runResult, 1) // Create a buffered channel
+		channels = append(channels, c)
+		go runTest(p, c)
+	}
+
+	wg.Wait()
```

Comment:
An important feature in the other languages is reporting intermediate results every second while the test is running.  Example output from .NET:

```
=== Test ===
    Current             Total          Average       CPU    WorkingSet    PrivateMemory
 25,195,915        25,195,915    25,119,342.75    25.15%        68.74M          161.21M
 24,333,189        49,529,104    24,603,047.21    25.82%        68.74M          161.33M
 23,544,658        73,073,762    24,433,749.18    26.07%        68.82M          161.47M

=== Results ===
Completed 73,073,762 operations in a weighted-average of 2.99s (24,433,749.18 ops/s, 0.000 s/op, 25.67% CPU)
```

Benefits:

1. Shows consistency of throughput over time, which can help identify insufficient warmup time, resource leaks, service throttling, etc.
2. Shows results in real-time while the test is running, without needing to wait the full duration.

Reply:
This is simple when only running a single thread, when it's run on multiple threads, is the `Total` column the total run over all threads and the `Current` is the value for a single thread?

I am working on CPU, WorkingSet, and PrivateMemory but I might add it in a later PR.

Reply:
All values are aggregated across all threads.  The threads need to either update shared variables (in a thread-safe way), or the main thread needs to query the threads and aggregate the values.

`Current` is the number of operations completed in the last second, `Total` is the cumulative operations completed since the run was started, and `Average` is `Total / Elapsed Time`.